plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererService.kt

```diff
@@ -825,11 +822,23 @@ data class RequestContext(
     val triggerTypeInfo: TriggerTypeInfo,
     val caretPosition: CaretPosition,
     val fileContextInfo: FileContextInfo,
-    val supplementalContext: SupplementalContextInfo?,
+    private val supplementalContextDeferred: Deferred<SupplementalContextResult>,
     val connection: ToolkitConnection?,
     val latencyContext: LatencyContext,
-    val customizationArn: String?
-)
+    val customizationArn: String?,
+) {
+    private var supplementalContext: SupplementalContextResult? = null
+
+    fun supplementalContext(): SupplementalContextResult = if (supplementalContext != null) {
+        supplementalContext as SupplementalContextResult
+    } else {
+        runBlocking {
```

Comment:
okk,that makes sense to me. I think I do it this way because we log the sup context early or on failure. But will update this part

Reply:
just think of one case, after awaiting and getting the real sup context response within a coroutine, we still need a data class to store pure `SupplemetalContextResult` instead of its deferred?

Reply:
to solve that, should we use 2 different data class of `RequestContext, one with Defer<SupplementalContext> and the other with real `SupplementalConext` ?

Reply:
oh richard i tried to restruct the flow and feel this should work 
https://github.com/aws/aws-toolkit-jetbrains/pull/4765/commits/0a29d95e1f199c1117b053e3a66096908377b07b

Reply:
should polish it a bit tho, wanna make `getRequestContext` fun signature still returns a regular object instead of defer but modify the call site